Mathlib/Combinatorics/SimpleGraph/Matching.lean

```diff
@@ -133,6 +133,53 @@ lemma IsMatching.coeSubgraph {G' : Subgraph G} {M : Subgraph G'.coe} (hM : M.IsM
   · obtain ⟨_, hw', hvw⟩ := (coeSubgraph_adj _ _ _).mp hy
     rw [← hw.2 ⟨y, hw'⟩ hvw]
 
+lemma IsMatching.exists_of_disjoint_sets_of_equiv {s t : Set V} (h : Disjoint s t)
+    (f : s ≃ t) (hadj : ∀ v : s, G.Adj v (f v)) :
+    ∃ M : Subgraph G, M.verts = s ∪ t ∧ M.IsMatching := by
+  haveI (v : V) : Decidable (v ∈ s) := Classical.dec _
+  use {
+    verts := s ∪ t
+    Adj := fun v w ↦ (if h : v ∈ s then f ⟨v, h⟩ = w else False) ∨
+          (if h : w ∈ s then f ⟨w, h⟩ = v else False)
+    adj_sub := by
+      intro v w h
+      simp only [dite_else_false] at h
+      cases' h with hl hr
+      · rw [← hl.choose_spec]
+        exact hadj ⟨v, hl.choose⟩
+      · rw [← hr.choose_spec, G.adj_comm]
+        exact hadj ⟨w, hr.choose⟩
+    edge_vert := by
+      intro v w hvw
+      simp only [dite_else_false] at hvw
+      cases' hvw with hl hr
+      · exact Set.mem_union_left t hl.choose
+      · rw [← hr.choose_spec]
+        exact Set.mem_union_right s (f _).coe_prop
+  }
```

Comment:
```suggestion
  use {
    verts := s ∪ t
    Adj := fun v w ↦ (∃ h : v ∈ s, f ⟨v, h⟩ = w) ∨ (∃ h : w ∈ s, f ⟨w, h⟩ = v)
    adj_sub := by
      intro v w h
      obtain (⟨hv, rfl⟩ | ⟨hw, rfl⟩) := h
      · exact hadj ⟨v, _⟩
      · exact (hadj ⟨w, _⟩).symm
    edge_vert := by aesop }
```
This proof simplifies quite a bit! The two ideas here are: you use the simp lemma `dite_else_false` a couple of times to simplify the adjacency relation, which suggests that the simplified version could make a nicer definition for it. Then instead of using the axiom of choice to extract the witness (in your lines 148-151), I used the `obtain` tactic to get it instead. And `obtain` can be combined with `cases'` to make that bit even shorter. Then `aesop` does the last goal!